### DISCUSSION TITLE
how to build this constraint for portfolio optimization

### DISCUSSION THREAD

**Discussion by s0ap** [cvxpy/cvxpy#1752] [Q&A] [answered]
Hi!

For portfolio optimization using cvxpy, how can I add active weight constraints? Active weight constraint here means i want to constraint weights something like this
benchmark_weight - 2% <= optimal_weight <= benchmark_weight + 2%

Can someone help me build this constraint?

Thanks!

**Comment by phschiele** [comment, marked as answer]
Hi @s0ap !

for simplicity, let's assume that the benchmark is an equal-weight portfolio:
```py
import numpy as np

n_securities = 10
benchmark_weights = np.ones(n_securities) / n_securities
```

Let's pick a minimum variance optimization as an example create a random covariance matrix as follows:
```py
np.random.seed(0)
A = np.random.rand(n_securities, n_securities)
covariance_matrix = A @ A.T
```

Finally, we can set up and solve the optimization problem:
```py
import cvxpy as cp

optimal_weights = cp.Variable(n_securities)
objective = cp.Minimize(cp.quad_form(optimal_weights, covariance_matrix))
constraints = [
    cp.sum(optimal_weights) == 1,
    benchmark_weights - 0.02 <= optimal_weights,
    optimal_weights <= benchmark_weights + 0.02
]
problem = cp.Problem(objective, constraints)
problem.solve()

print(optimal_weights.value)
>>> [0.08 0.08 0.08 0.08 0.12 0.12 0.12 0.08 0.12 0.12]
```
Notice how similar the constraint syntax is to your description!

**Reply by s0ap** [reply]
I see, so for vector scalar operations, we do not need to use special cvx methods (like cp.sum) - many thanks!
